snapcraft/commands/core22/upload.py

```diff
@@ -106,6 +155,54 @@ def run(self, parsed_args):
         emit.message(message)
 
 
+def _validate_components(
+    provided_components: list[ComponentOption], snap_yaml: dict[str, Any]
+) -> None:
+    """Validate component data.
+
+    Validates:
+     - The components provided by the user match the components listed in the snap's metadata.
+     - All component files exist.
+
+    :param provided_components: A list of ComponentOptions provided by the user.
+    :param snap_yaml: The snap's metadata.
+    """
+    # get the names of the components provided by the user
+    provided_names = {
+        component.name for component in provided_components if component.name
+    }
+    emit.debug(f"Components provided by the user: {provided_names}")
+
+    if snap_yaml.get("components"):
```

Comment:
`"components" in snap_yaml` could be marginally better here.

Reply:
Updated